csharp/src/SeedLang.Block/BxfObject.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Serialization;
+using System.Collections.Generic;
+
+namespace SeedLang.Block {
+  // The intermediate data structure for serializing or parsing a SeedBlock module.
+  //
+  // BxfWriter uses the following types to prepare in-memory objects then serializes them to JSON
+  // format. BxfReader reads JSON format then stores the parsed result with the following types as
+  // an intermediate structure.
+  //
+  // TODO: It's possible to build an auto-generator to create the following C# code based on
+  // /schemas/bxf.schema.json. Consider this possibility in the future.
+  [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore,
+              NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+  internal class BxfObject {
+    public string Schema = BxfConstants.Schema;
+    public string Version = BxfConstants.Version;
+    public string Author = null;
+    public string Copyright = null;
+    public string License = null;
+    public BxfModule Module = new BxfModule();
+  }
+
+  [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore,
+              NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+  internal class BxfModule {
+    public string Name = null;
+    public string Doc = null;
+    public List<BxfBlock> Blocks = new List<BxfBlock>();
+  }
+
+  [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore,
+              NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+  internal class BxfBlock {
+    public string Id = null;
+    public string Type = null;
+    public string Doc = null;
+    public string Content = null;
```

Comment:
Yes, the exported blocks are always in a flat list, without keeping their tree structure.

Option 2 doesn't work for this case:

The if ... else ... block has 2 compound statements and each of them has one or more single statements. It it a tree structure even when all its sub-blocks are single blocks. E.g.,

if (a > 3) {
   b = 1;
   c = 2;
} else {
   b = 3;
   c = 4;
}

It's not possible to use a flat list to hold all the 4 child statement: "b=1", "c=2", "b=3" and "c=4".  Otherwise, we don't know what is the boundary between the if ... part and the else ... part.

That's why I suggested defining both "child statement dock type" and "next statement dock type".

Thus, the if...else... block will only have two direct child statements: "b=1" and "b=3". These two statements can be stored in a flat list, and be serialized and parsed as you suggested.

"c=2" will be docked to "b=1" as its NextStatement, "c=4" will be docked to "b=3" as its statements. "c=2" and "c=4" will be serialized and parsed separately.

Does the above example answer your question?

Reply:
For the above example, the tree might be:
```
"if-else statement" -----> "child statement b=1" ------> "next statement c=2"
         |
         -----------> "child statement b=3" --------> "next statement c=4"
```
I was just thinking if this tree could be serialized to a JSON string directly without storing to a BxfObject first. Maybe it's not possible, because JSON strings can only store List and Dictionary, and cannot express this kind of tree structure.
Let's keep your current design.